### ISSUE TITLE
JVM crash on Ubuntu 21

### ISSUE THREAD

**Issue by sheidrich-inet** [jnr/jffi#120]
Hello,

while testing our application on a new Ubuntu 21 system, a JVM crash always occurred during startup. 
Upon closer analysis we saw that this crash is in the native code. 

We are trying to read in an image with OpenJpeg at this point. The code looks something like this:

```
        System.setProperty( "jnr.ffi.library.path", "<path>" );
       
        // to get a better message and exception before the registration
        new OpenJpeg();

        // register in ImageIO
        IIORegistry registry = IIORegistry.getDefaultInstance();
        registry.registerServiceProvider( new OpenJp2ImageReaderSpi() );
        registry.registerServiceProvider( new OpenJp2ImageWriterSpi() );

        // test if it is working
        try (InputStream testLogo = getClass().getResourceAsStream( "verification.jp2" ) ){
            BufferedImage testImage = ImageIO.read( testLogo );
            if( testImage == null ) {
                throw new IOException( "OpenJpeg library not initalized correctly, Jpeg2000 images will not be displayed." );
            }
        } catch (Throwable e) {
		e.printStackTrace();
	}

```
The crash happened while reading the image with 
```
BufferedImage testImage = ImageIO.read( testLogo );
```
I upload the error log file and hope you can tell me, whats the problem is and how to solve it. With an older Ubuntu this works.

[hs_err_pid7642.log](https://github.com/jnr/jffi/files/7788262/hs_err_pid7642.log)

Best regards,

S. Heidrich

**Comment by Horcrux7** [comment]
This is a crash log with `System.setProperty( "jnr.ffi.asm.enabled", "false" );`
[hs_err_pid2558.log](https://github.com/jnr/jffi/files/7792379/hs_err_pid2558.log)

It looks a little different. We tried it, hoping to find a workaround and the cause of the crash.

It is also possible to set a breakpoint at the last non-native method before the crash. So it is not a timing or threading issue.

**Comment by Horcrux7** [comment]
**Possible Workarounds**
* Looking for a workaround, we scaled down our application (removing many optional plugins/jar files) and that solved the problem.
* If the native code was initialized before the first full GC, it works without VM crash.
* If we calling `System.gc()` in our code it crash also with Java 17 which has work before.
* But in a simple test program a `System.gc()` does not force the crash.
* Only calling `StubLoader.isLoaded();` at a very early stage does not help. Only the early init of the [imageio-openjpeg](https://github.com/dbmdz/imageio-jnr) library seems to solve the problem.

If we can present any helpful information then tell us how we can do this.

**Comment by headius** [comment]
I missed this over the holiday. Thank you for the report!

Does this reproduce consistently? I am not an expect on debugging native code but if I can reproduce there's a good chance I can find the problem.

**Comment by Horcrux7** [comment]
Yes, it was reproducible consistently. We have try it on two Ubuntu 21 systems and it crash ever on every run.

Because this not occur with a simple test case you can download our SDK at https://download.inetsoftware.de/clear-reports-sdk-latest.zip
And run the start script `i-net Clear Reports\startDesigner.sh` or `i-net Clear Reports\startServer.sh`. After 30-60 seconds it will crash.

The jnr files are in the plugin `i-net Clear Reports\plugins\jpeg2000.zip`. You can replace the jar files in this plugin with debug versions or newer versions if you want.

Or for simpler debugging in an IDE you can set the Java command line switch `-Djpeg2000.classpath=<path to classes>` (without slash at path end)

**Comment by gamma** [comment]
I can now reproduce the issue with an `eclipse-temurin/17-jre-alpine` Docker image as well.

Are there any updates on this issue?

**Comment by gamma** [comment]
Here is another update: the crash always occurs with eclipse-temurin Java VMs, I checked v11 to v17. The Adoptium 11 Java VM does not crash, so there may be an issue with either them or this library.

**Comment by gamma** [comment]
Here is another update. I created an issue with the Temurin support at: https://github.com/adoptium/adoptium-support/issues/477 - you can use the sample code there as well.

**Comment by headius** [comment]
I will try to reproduce today and see if I can find a problem.

**Comment by Horcrux7** [comment]
The new crash is related to a incompatibility of `glibc` and `musl`. Alpine use musl and the native library was compiled with glibc. If there can be a helpful exception would be a great improvement if this is possible. A crash is very bad.

Details can you find at: https://github.com/dbmdz/imageio-jnr/issues/192

The old crash with Ubuntu 21 have we solve with switching to Java 17. You can close this issue if there is no improvement possible in such a case.
